includes/Modules/Analytics/Advanced_Tracking.php

```diff
@@ -0,0 +1,175 @@
+
+	/**
+	 * List of event configurations to be tracked.
+	 *
+	 * @since n.e.x.t.
+	 * @var array
+	 */
+	private $event_configurations;
+
+	/**
+	 * Main class plugin detector instance.
+	 *
+	 * @since n.e.x.t.
+	 * @var Plugin_Detector
+	 */
+	private $plugin_detector;
+
+	/**
+	 * Advanced_Tracking constructor.
+	 *
+	 * @since n.e.x.t.
+	 *
+	 * @param Plugin_Detector $mock_plugin_detector optional plugin detector used for testing.
+	 */
+	public function __construct( $mock_plugin_detector = null ) {
+		$this->supported_plugins = array(
+			'Contact Form 7'   => array(
+				'check_name'        => 'WPCF7_PLUGIN_DIR',
+				'check_type'        => Plugin_Detector::TYPE_CONSTANT,
+				'event_config_list' => new CF7_Event_List(),
+			),
+			'Formidable Forms' => array(
+				'check_name'        => 'load_formidable_forms',
+				'check_type'        => Plugin_Detector::TYPE_FUNCTION,
+				'event_config_list' => new FormidableForms_Event_List(),
+			),
+			'Ninja Forms'      => array(
+				'check_name'        => 'NF_PLUGIN_DIR',
+				'check_type'        => Plugin_Detector::TYPE_CONSTANT,
+				'event_config_list' => new NinjaForms_Event_List(),
+			),
+			'WooCommerce'      => array(
+				'check_name'        => 'WC_PLUGIN_FILE',
+				'check_type'        => Plugin_Detector::TYPE_CONSTANT,
+				'event_config_list' => new Woocommerce_Event_List(),
+			),
+			'WPForms'          => array(
+				'check_name'        => 'WPFORMS_PLUGIN_DIR',
+				'check_type'        => Plugin_Detector::TYPE_CONSTANT,
+				'event_config_list' => new WPForms_Event_List(),
+			),
+		);
+		if ( null === $mock_plugin_detector ) {
+			$this->plugin_detector = new Plugin_Detector( $this->supported_plugins );
+		} else {
+			$this->plugin_detector = $mock_plugin_detector;
+		}
+	}
+
+	/**
+	 * Registers functionality through WordPress hooks.
+	 *
+	 * @since n.e.x.t.
+	 */
+	public function register() {
+		add_action(
+			'wp_enqueue_scripts',
+			function() {
+				$this->set_up_advanced_tracking();
+			},
+			11
+		);
+		add_filter(
+			'googlesitekit_amp_gtag_opt',
+			function( $gtag_amp_opt ) {
+				return $this->set_up_advanced_tracking_amp( $gtag_amp_opt );
+			}
+		);
+	}
+
+	/**
+	 * Injects javascript to track active events.
+	 *
+	 * @since n.e.x.t.
+	 */
+	private function set_up_advanced_tracking() {
+		if ( ! wp_script_is( 'google_gtagjs' ) ) {
+			return;
+		}
+		$this->configure_events();
+		( new Measurement_Code_Injector( $this->event_configurations ) )->inject_event_tracking();
+	}
+
+	/**
+	 * Adds triggers to AMP configuration.
+	 *
+	 * @since n.e.x.t.
+	 *
+	 * @param array $gtag_amp_opt gtag config options for AMP.
+	 * @return array $gtag_amp_opt gtag config options for AMP.
+	 */
+	private function set_up_advanced_tracking_amp( $gtag_amp_opt ) {
+		$this->configure_events();
+		foreach ( $this->event_configurations as $event_config ) {
+			$gtag_amp_opt['triggers'][ $event_config->get_amp_trigger_name() ] = $event_config->to_amp_config();
+		}
```

Comment:
It would be a bit safer to add a check above the `foreach`, to set `$gtag_amp_opt['triggers']` to an empty array in case it's not set yet.

Reply:
This response is really to the comment above, but I am unable to add a response to that comment for some reason. 

Are you suggesting that the 'event_config_list' key in this array only contain the name of the class in string form rather than the actual instance? If that is the case, I think we would need the Measurement_Event_Factory class again to actually instantiate the lists.

Reply:
Won't this still instantiate the supported plugin configurations even if they are not active and we don't need them? A factory pattern could produce them dynamically and on demand.